### ISSUE TITLE
misc.py Bug Report

### ISSUE THREAD

**Issue by elmo1032** [elmo1032/BitDelta#17]
I found several potential issues in the provided code:

1. In the `find_corr_stddev` function, the `corr` variable might not be calculated correctly due to the missing `to("cpu")` conversion for `finetuned_weight` in the calculation of `covar`.
2. The `corr` variable might cause an error if its value is exactly 1, as the `stddev` calculation will receive a tensor with a single value of 0, which cannot be converted to a float using the `.item()` method.
3. In the `MixtralBinaryDiff` class, the `nn.SiLU()` function call is missing parentheses to correctly call the function.
4. In the `compress_mixtral_moe_diff` function, the `new_experts` list is created without calling `nn.ModuleList()` with the `new_experts` argument.
5. The `compress_mixtral_moe_diff` function uses `setattr` to set the `experts` attribute of the `module` to `None`, but it should be set to an empty `nn.ModuleList()`.
6. The `compress_mixtral_moe_diff` function does not call `gc.collect()` and `torch.cuda.empty_cache()` after setting the `experts` attribute to the new list.
7. In the `dequantize_model` function, the `dequantize_8bit` and `dequantize_4bit` functions do not handle the case when the input layer is not an instance of `Linear8bitLt` or `QuantLinear`.
8. The `dequantize_model` function does not check if the `quantized_model` has the same architecture as the `model`.
9. The `dequantize_model` function does not handle the case when the `quant_type` is not "8bit" or "4bit".
10. The `LoRADiff` class does not handle the case when the `base` and `finetune` matrices have different shapes.
11. The `LoRADiff` class does not handle the case when the `rank` or `niter` arguments are not provided during initialization.
12. The `LoRADiff` class does not define a `forward` method to compute the output of the layer.

Here's the corrected code with the identified issues addressed:

```python
import torch
import torch.nn as nn
import gc
from bitdelta.diff import BinaryDiff

def find_corr_stddev(base_model, finetuned_model):
    corrs = []
    stddevs = []  # store std dev of base weight - finetuned weight
    for name, module in base_model.named_modules():
        if "mlp" in name or "self_attn" in name:
            for subname, submodule in module.named_children():
                if "proj" in subname:
                    base_weight = base_model.get_submodule(name + "." + subname).weight.view(-1).to("cpu")
                    finetuned_weight = finetuned_model.get_submodule(name + "." + subname).weight.view(-1).to("cpu")
                    # find correlation, assuming unravelled
                    base_stddev = torch.std(base_weight).float()
                    finetuned_stddev = torch.std(finetuned_weight).float()

                    covar = torch.dot(base_weight, finetuned_weight).float() / base_weight.shape[0]
                    corr = covar / (base_stddev * finetuned_stddev)

                    if corr.item() >= 1:
                        stddev = torch.tensor(0).float()
                    else:
                        stddev = torch.sqrt(base_stddev**2 + finetuned_stddev**2 - 2 * corr * base_stddev * finetuned_stddev)
                    corrs.append(corr.item())

                    stddevs.append(stddev.item())

    return corrs, stddevs

class MixtralBinaryDiff(nn.Module):
    def __init__(self, w1, mean_w1, w2, mean_w2, w3, mean_w3):
        super().__init__()
        self.w1 = BinaryDiff(mean_w1, w1)
        self.w2 = BinaryDiff(mean_w2, w2)
        self.w3 = BinaryDiff(mean_w3, w3)
        self.act_fn = nn.SiLU()()  # Added parentheses to correctly call the function

    def forward(self, hidden_states):
        current_hidden_states = self.act_fn(self.w1(hidden_states)) * self.w3(hidden_states)
        current_hidden_states = self.w2(current_hidden_states)
        return current_hidden_states

def compress_mixtral_moe_diff(model):
    for name, module in model.named_modules():
        if name.endswith("block_sparse_moe"):
            experts = module.experts
            new_experts = nn.ModuleList()  # Added nn.ModuleList()
            mean_w1 = torch.mean(torch.stack([expert.w1.weight for expert in experts]), dim=0).detach()
            mean_w2 = torch.mean(torch.stack([expert.w2.weight for expert in experts]), dim=0).detach()
            mean_w3 = torch.mean(torch.stack([expert.w3.weight for expert in experts]), dim=0).detach()
            for expert in experts:
                w1 = expert.w1.weight
                w2 = expert.w2.weight
                w3 = expert.w3.weight
                new_expert = MixtralBinaryDiff(w1, mean_w1, w2, mean_w2, w3, mean_w3)
                new_experts.append(new_expert)

            del experts, mean_w1, mean_w2, mean_w3
            setattr(module, "experts", nn.
